# Morse grid imports

Here we look at applied examples to import data for use in Dr Peter Morse 3D visualtsation software _(in prep.)_. Data must be exported as $1800 \times 3600$ px png files. This might change in future. 

Generally, the code in this tutorial runs rather slow as there is a large amount of interpolation. Have patience. 

Folder `../local` is used to download data and export pngs. 

In [ ]:
#import warnings
#warnings.filterwarnings('ignore')

#config_file = "../agrid/agrid.py"
#with open(config_file) as f:
#    code = compile(f.read(), config_file, 'exec')
#    exec(code, globals(), locals())

# ... or use you Python path or present working directory:
import sys
sys.path.append('..')
from agrid.agrid import *



import vtk
from vtk.util.numpy_support import vtk_to_numpy

In [ ]:
from agrid.features import export_morse_png

## SMEAN

Download the mantel model from from Becker and Boschi, The University of Texas at Austin. See original paper [Becker and_Boschi 2002](http://www-udc.ig.utexas.edu/external/becker/becker_and_boschi_2002.pdf) for details.

In [ ]:
! mkdir -p ../../data/smean
! wget -nc http://www-udc.ig.utexas.edu/external/becker/ftp/smean_grd.tgz \
    -O ../../data/smean/smean_grd.tgz
! tar -xvzf ../../data/smean/smean_grd.tgz -C ../../data/


In this example, the data is global and we define a grid with the right rsolution from start. However, as we will see later, we could work on regional data as well. The exported png will always have a global coverage with alpha covering the excluded areas. 

We read depth values for smean from a provided ascii file. However we also have a regular depth dimention, defined as regular 50km depths. 

In [ ]:
#columns from left to right represent radius,depth,density,Vpv,Vph,Vsv,Vsh,eta,Q-mu,Q-kappa




smean_depths = km * np.loadtxt('../../data/smean/depths.dat')[::-1]
max_depth = np.max(smean_depths)

depths = km * np.array([*range(0, 1000, 20)] + [*range(1000, int(max_depth)//km, 50)])

prem_d, prem_vsv, prem_vsh = km * np.loadtxt('../../../data/models/PREM_1s.csv', 
                              delimiter = ',', 
                              usecols = [1, 5, 6],
                              unpack=True)

prem_vsv = prem_vsv[prem_d < max_depth]
prem_vsh = prem_vsh[prem_d < max_depth]
prem_d = prem_d[prem_d < max_depth]

print(depths)

In [ ]:
world = Grid(crs_tgt = 4326, 
             left= -180, up=90, down= -90, right=180.0, 
             res=(0.099999999999,0.099999999999), 
             depths = depths)

array = np.empty((world.ny, world.nx, world.ds.coords['Z'].size))
array[:] = np.nan
print(world.ds)

In [ ]:
world.ds['PREM_VSV'] = (('Z'), world.change_coord(prem_vsv, prem_d, 'Z') )

Now, we can read SMEAN files. set_center alows us to read 0-360 as -180 - 180 lon. read_grid uses default nearest neighbour interpolation of datapoints to fit the predefined Morse grid. 

In [ ]:
smean_array = np.empty((world.ny, world.nx, smean_depths.size))

for i, a in enumerate(smean_depths):
    print(a//km, end=', ')
    index_name = smean_depths//km - i # Get right file name for each depth
    fname = '../../data/smean/dvs.%s.grd'%(i+1)
    if os.path.isfile(fname):
        smean_array[:,:,i] = world.read_grid(fname, xyz = ('lon','lat','z'), set_center = True)
        
#array /= 100    #"The grid file values are given are percentage v_S wave 
#                #speed variation related to PREM, with layer averages removed."
        
#world.ds['SMEAN'] = (('Y', 'X', 'Z'), array)

Finally, we export the 3D grids as png files for each depth slice. We store a set of original depth slices in ''morse/smean/z_orig/' and also the interpolated depth slices in 'morse/smean/z_even/'. This might take some time. 

In [8]:
smean_min = np.nanmin(smean_array)
smean_max = np.nanmax(smean_array)

print(smean_min, smean_max)

-8.248664885133621 6.73640155598332


In [10]:
new_smean_array  = world.change_coord(smean_array, old=smean_depths, new='Z')

In [11]:
np.shape(new_smean_array)

(1799, 3599, 87)

In [12]:
world.ds['SMEAN'] = (('Y', 'X', 'Z'), new_smean_array)

In [15]:
import pickle
pickle.dump(world, open( "../local/world.p", "wb" ), protocol=4)
#world = pickle.load( open( "../local/world.p", "rb" ) )





OSError: [Errno 22] Invalid argument

In [17]:
! mkdir -p ../local/morse/smean

In [23]:
for i, z in enumerate(world.ds['Z'][::-1]):
    report = export_morse_png(world,
                        data = world.ds['SMEAN'].isel(Z=i).values, 
                        png_name = '../local/morse/smean_RGB/%04d_smean.png'%(int(z)//km), 
                        v_min = -9, v_max = 9, 
                        png_format = 'RGB',
                        set_geometry = False)
    print(report)
    with open("../local/morse/smean_RGB/log_smean_extra.txt", "a") as log_file:
        log_file.write(report)
    

../local/morse/smean_RGB/2800_smean.pnggreyscale: False
alpha: False
planes: 3
bitdepth: 8
interlace: 0
size: (3599, 1799)
../local/morse/smean_RGB/2800_smean.png 
min v: -9 max v: 9 bit depth: 8
bands: (1799, 3599, 3) interpolation: nearest
data 	  norm 	  to png 	 png 
nan 	  0.50 	 127.50 	  127 
nan 	  0.50 	 127.50 	  127 
 

../local/morse/smean_RGB/2750_smean.pnggreyscale: False
alpha: False
planes: 3
bitdepth: 8
interlace: 0
size: (3599, 1799)
../local/morse/smean_RGB/2750_smean.png 
min v: -9 max v: 9 bit depth: 8
bands: (1799, 3599, 3) interpolation: nearest
data 	  norm 	  to png 	 png 
nan 	  0.50 	 127.50 	  127 
nan 	  0.50 	 127.50 	  127 
 

../local/morse/smean_RGB/2700_smean.pnggreyscale: False
alpha: False
planes: 3
bitdepth: 8
interlace: 0
size: (3599, 1799)
../local/morse/smean_RGB/2700_smean.png 
min v: -9 max v: 9 bit depth: 8
bands: (1799, 3599, 3) interpolation: nearest
data 	  norm 	  to png 	 png 
nan 	  0.50 	 127.50 	  127 
nan 	  0.50 	 127.50 	  127 
 

.

../local/morse/smean_RGB/1550_smean.pnggreyscale: False
alpha: False
planes: 3
bitdepth: 8
interlace: 0
size: (3599, 1799)
../local/morse/smean_RGB/1550_smean.png 
min v: -9 max v: 9 bit depth: 8
bands: (1799, 3599, 3) interpolation: nearest
data 	  norm 	  to png 	 png 
-1.72 	  0.40 	 103.15 	  103 
1.02 	  0.56 	 141.92 	  141 
 

../local/morse/smean_RGB/1500_smean.pnggreyscale: False
alpha: False
planes: 3
bitdepth: 8
interlace: 0
size: (3599, 1799)
../local/morse/smean_RGB/1500_smean.png 
min v: -9 max v: 9 bit depth: 8
bands: (1799, 3599, 3) interpolation: nearest
data 	  norm 	  to png 	 png 
-1.74 	  0.40 	 102.85 	  102 
1.03 	  0.56 	 142.07 	  142 
 

../local/morse/smean_RGB/1450_smean.pnggreyscale: False
alpha: False
planes: 3
bitdepth: 8
interlace: 0
size: (3599, 1799)
../local/morse/smean_RGB/1450_smean.png 
min v: -9 max v: 9 bit depth: 8
bands: (1799, 3599, 3) interpolation: nearest
data 	  norm 	  to png 	 png 
-1.76 	  0.40 	 102.52 	  102 
1.04 	  0.56 	 142.24 	  

../local/morse/smean_RGB/0720_smean.pnggreyscale: False
alpha: False
planes: 3
bitdepth: 8
interlace: 0
size: (3599, 1799)
../local/morse/smean_RGB/0720_smean.png 
min v: -9 max v: 9 bit depth: 8
bands: (1799, 3599, 3) interpolation: nearest
data 	  norm 	  to png 	 png 
-1.39 	  0.42 	 107.76 	  107 
1.03 	  0.56 	 142.14 	  142 
 

../local/morse/smean_RGB/0700_smean.pnggreyscale: False
alpha: False
planes: 3
bitdepth: 8
interlace: 0
size: (3599, 1799)
../local/morse/smean_RGB/0700_smean.png 
min v: -9 max v: 9 bit depth: 8
bands: (1799, 3599, 3) interpolation: nearest
data 	  norm 	  to png 	 png 
-1.39 	  0.42 	 107.76 	  107 
1.07 	  0.56 	 142.65 	  142 
 

../local/morse/smean_RGB/0680_smean.pnggreyscale: False
alpha: False
planes: 3
bitdepth: 8
interlace: 0
size: (3599, 1799)
../local/morse/smean_RGB/0680_smean.png 
min v: -9 max v: 9 bit depth: 8
bands: (1799, 3599, 3) interpolation: nearest
data 	  norm 	  to png 	 png 
-1.37 	  0.42 	 108.06 	  108 
1.06 	  0.56 	 142.58 	  

../local/morse/smean_RGB/0220_smean.pnggreyscale: False
alpha: False
planes: 3
bitdepth: 8
interlace: 0
size: (3599, 1799)
../local/morse/smean_RGB/0220_smean.png 
min v: -9 max v: 9 bit depth: 8
bands: (1799, 3599, 3) interpolation: nearest
data 	  norm 	  to png 	 png 
-1.44 	  0.42 	 107.16 	  107 
1.74 	  0.60 	 152.14 	  152 
 

../local/morse/smean_RGB/0200_smean.pnggreyscale: False
alpha: False
planes: 3
bitdepth: 8
interlace: 0
size: (3599, 1799)
../local/morse/smean_RGB/0200_smean.png 
min v: -9 max v: 9 bit depth: 8
bands: (1799, 3599, 3) interpolation: nearest
data 	  norm 	  to png 	 png 
-1.40 	  0.42 	 107.64 	  107 
1.71 	  0.59 	 151.67 	  151 
 

../local/morse/smean_RGB/0180_smean.pnggreyscale: False
alpha: False
planes: 3
bitdepth: 8
interlace: 0
size: (3599, 1799)
../local/morse/smean_RGB/0180_smean.png 
min v: -9 max v: 9 bit depth: 8
bands: (1799, 3599, 3) interpolation: nearest
data 	  norm 	  to png 	 png 
-1.38 	  0.42 	 107.96 	  107 
1.94 	  0.61 	 154.98 	  

## GLAD M15

Now let's import the model from [Bozdağ et al (2016)](https://academic.oup.com/gji/article/207/3/1739/2404568). Unfortunately, it appears that the data is not availible from any open repo (?!), but I belive that the group is willing to assist if contacted. 

The SMEAN data was easy to import, but exporting the GLAD-M15 data is more compicated as I only get access to an unstructured grid, in a normalised space. 

In [ ]:
# load a vtk file to reader
reader = vtk.vtkXMLUnstructuredGridReader()
reader.SetFileName('../local/GLAD-M15/reg_1_dvsv.vtu')
reader.Update()

# Get the coordinates of nodes in the mesh
nodes_vtk_array= reader.GetOutput().GetPoints().GetData()

#The data field is the first scalar in the vtu file
data_vtk_array = reader.GetOutput().GetPointData().GetArray(0)

print(data_vtk_array, nodes_vtk_array)

We get two arrays. Note that sizes are equal (Number Of Tuples: 4712064 and Number Of Tuples: 4712064) and the coordinate are 3D (NumberOfComponents: 3). Read coordinates and data as numpy arrays and have a look: 

In [ ]:
nodes_numpy_array = vtk_to_numpy(nodes_vtk_array)
X, Y, Z = nodes_numpy_array[:,0] , nodes_numpy_array[:,1] , nodes_numpy_array[:,2]
V = vtk_to_numpy(data_vtk_array)

print(X[:10],Y[:10],Z[:10],V[:10], sep='\n\n')

The coordinates are stored in the range [-1..1]. We assume a spherical Earth and compute an array with every points distance from centre. Function `sphere_to_layer()` takes a slice from the sphere and save to a 2D array. delta_r should be as small as possible, for best vertical resolution, but if too few points are used the vertical resolution gets bad. 

We loop through the harmonics of the sphere and read all points within a spherical shell defined by the radius +/- delta_r

In [ ]:
def sphere_to_layer(d, R, V, LAT, LON, xxx, yyy, 
                    delta_r = 32*km, 
                    min_s = 5000,
                    r_earth =  (6357*km + 6378*km) / 2, 
                    interpolation = 'nearest'):
    
    A = np.zeros_like(xxx).astype('float')
    A[:] = np.nan
    upper = 1-(d-delta_r)/r_earth
    lower = 1-(d+delta_r)/r_earth
    S = (R > lower) & (R < upper) # Select points in spherical shell
    s_sum = np.count_nonzero(S) # Check how many points in shell

    if s_sum >= min_s:
        A = interpolate.griddata((LON[S], LAT[S]),
                V[S], (xxx, yyy), method = interpolation)
    print(d//1000, 'km \t N=', sum(S),u'\t \N{BLACK RIGHT-POINTING TRIANGLE}', np.nanmean(A))
    return np.flipud(A) #Because ulike lat, rows start from top. 

R = np.sqrt(X*X + Y*Y + Z*Z) # The distance from each point to centre of spherical Earth. 

LAT = world.shape3[0]/180 * (np.arccos( Z / R) * 180/np.pi) 
LON = world.shape3[1]/360 * (np.arctan2(Y, X) * 180/np.pi + 180)

xxx, yyy = np.meshgrid(range(0, world.shape3[1]), range(world.shape3[0], 0, -1)) 

glad = np.zeros(world.shape3) # self.shape3 is a tuple of the models dimensions

for i, d in enumerate(world.ds['Z'].values):
    glad[:,:,i] = sphere_to_layer(d, R, V, LAT, LON, xxx, yyy)

Unfortunately there are not enugh data points at 1350, 1450, 2150 and 2250 km depth. We'd need to look in a broad spherical shell to get datapoints: 

In [ ]:
missing_i, missing_d = [], []

for i in range(glad.shape[2]):
    if (np.count_nonzero(np.isnan(glad[:,:,i]) ) > 1000):
        missing_i.append(i)
        missing_d.append(world.ds['Z'].values[i])
      
for i,d in zip(missing_i, missing_d):
    print(i, end=' ')
    glad[:,:,i] = sphere_to_layer(d, R, V, LAT, LON, xxx, yyy, delta_r = 45*km)    

However, we can be more precise with the first layer. 

In [ ]:
glad[:,:,0] = sphere_to_layer(50*km, R, V, LAT, LON, xxx, yyy, delta_r = 14*km)   
glad[:,:,1] = sphere_to_layer(150*km, R, V, LAT, LON, xxx, yyy, delta_r = 20*km)   

We save the array to the object and have a look. 

In [ ]:
world.ds['GLAD'] = (('Y', 'X', 'Z'), glad)
glad_min = np.nanmin(world.ds['GLAD'].values)
glad_max = np.nanmax(world.ds['GLAD'].values)

print(glad_min, glad_max)
#print(smean_min, smean_max)

z_map = 0
world.map_grid(world.ds['GLAD'].isel(Z=z_map), cmap='magma_r', vmin=-0.06, vmax=0.06)
#world.map_grid(world.ds['SMEAN'].isel(Z=z_map), cmap='magma_r', vmin=-0.06, vmax=0.06)

In [ ]:
world.ds['DIFF'] = world.ds['GLAD']-world.ds['SMEAN']

diff_min = np.nanmin(world.ds['DIFF'].values)
diff_max = np.nanmax(world.ds['DIFF'].values)

print(diff_min, diff_max)

world.map_grid(world.ds['DIFF'].isel(Z=z_map), cmap='BrBG', vmin=-0.06, vmax=0.06)

In [ ]:
! mkdir -p ../local/glad/16_bit

In [ ]:
for i, z in enumerate(world.ds['Z']):
    report = export_morse_png(world, world.ds['GLAD'].isel(Z=i).values, '../local/glad/16_bit/%04d_%s_glad.png'%(int(z)//km, i+1), 
                          v_min = glad_min, v_max = glad_max, 
                              set_geometry = False, 
                             png_format = 'RGB', 
                             bit_depth = 16)
    with open("../local/morse/glad/log_16bit.txt", "a") as log_file:
        log_file.write(report)
    print(report)
    

Save the grid: 

In [ ]:
world.save(world.ds, '../local/smean_and_glad.nc')

We could also interpolate new depth values to make a regular 3D grid. 

In [ ]:
world.ds['Z_NEW'] = range(0,2850*km, 100*km)

world.ds['SMEAN_INTER'] = ( ('Y', 'X', 'Z'), 
                           world.change_coord(world.ds['SMEAN'], world.ds['Z'], world.ds['Z_NEW']) )

In [ ]:
world.ds

# AuSREM

Australian crustal and lithospheric seismic model.

Can be downloaded from http://rses.anu.edu.au/seismology/AuSREM/Downloads/ We assume a local file: 

In [ ]:
fname = '../local/ausrem_SV_100.txt'
data = world.read_ascii(fname)

Make a mask, as the 'nearest neigbur' interpolation tested, will extrapolate values to all cells.

In [ ]:
alpha_master = np.isfinite(data)

plt.imshow(alpha_master)
plt.show()

Here, we test a number of different formats for the output png file: 

In [ ]:
for _format in ['L', 'LA', 'RGB', 'RGBA']:
    for _bit_depth in [8, 16]:
        for _interpol in ['nearest', 'linear', 'cubic']:
            data = world.read_ascii(fname, interpol =_interpol)
            
            if _interpol == 'nearest':
                how_confine = 'mask'
                mask_to_value = np.nan
            else:
                how_confine = 'input'
                mask_to_value = None
            
            report = export_morse_png(world,
                          data = data, 
                          png_format = _format,
                          png_name = '../local/AuSREM/AuSREM_100km_%s_%sbit_%s.png'%( _interpol,_bit_depth,_format),
                          interpol_method =  _interpol,
                          confine_data = how_confine,
                          confine_mask = alpha_master,
                        mask_to_value = mask_to_value,
                          v_min = 4, v_max = 5, 
                          clip=True,
                          set_geometry = False, 
                          bit_depth=_bit_depth)

            print(report)
            with open("../local/AuSREM_log.txt", "a") as log_file:
                log_file.write(report)

In [ ]:
!sips -g all ../local/AuSREM/AuSREM_100km_nearest_16bit_L.png
#!open ../local/new/AuSREM_100km_nearest_16bit_L.png